datalad/downloaders/providers.py

```diff
@@ -173,28 +177,37 @@ def __iter__(self):
 
     @classmethod
     def from_config_files(cls, files=None, reload=False):
```

Comment:
note that ideally there should be a dedicated test which would verify that it finds present files and loads then in desired (here it is not actually, system is added last) order:  datalad -> dataset -> system -> user (I think this would be the best, from hardcoded, to shipped, to custom local); so that settings in e.g. dataset overload the one in datalad. 
 Would be nice to test in user directory overload as well, but HOME might be left not modified to custom one for  some test runs... not sure if appdirs has "memory" but you could create a temp directory and `patch.dict('os.environ', {'HOME': path})` to mock for that specific test 
makes sense or I talk jibberish?

Reply:
I'll see about adding some tests, but I'm not sure how to fake the system config files in a testing context or how to make the user config test work in a cross platform way that wouldn't end up just using the same AppDir config (implying that it would be a somewhat tautologically and not doing much useful testing.)

Also, shouldn't the precedence be datalad dist -> system -> user -> dataset? (in any case you're right about the fact that system taking precedence is definitely wrong..)

Reply:
may be you are right.  my thinking was -- what if dataset is old and provided settings are no longer valid and need adjustment... well, dataset could be fixed up I guess.
@mih blind me can't find a clear description in the docs about our order in the `config` handling... I just remember that env variable overrides anything, I guess that it is system -> user -> dataset ->  dataset local (.git/config) -> env , right?

Reply:
Nevermind, I just realized I can fake it by patching dirs.user_config_dir and dirs.site_config_dir for the test...

Reply:
> blind me can't find a clear description in the docs about our order in the config handling... I just remember that env variable overrides anything, I guess that it is system -> user -> dataset -> dataset local (.git/config) -> env , right?

Not a specific outline of the order, but the [configuration docs](http://docs.datalad.org/en/docs/config.html) claim that "global and local settings on a particular machine always override configuration shipped with a dataset", so I'd guess your guess isn't right (or at least isn't supposed to be).

Reply:
That documentation seems to specifically be referring to the git configuration settings (where that makes some sense, since those settings are intended for user customization of settings). As an outsider, it makes less sense to me for datalad downloader configuration, because it means the system (if you call that "global") settings will always override anything thing set in a data set, and there would be no way to customize things on a per dataset level on any system where a system administrator has set something up.

But I can go with that if that's the intended design of datalad.

Reply:
@driusan Nothing is set in stone. We are running with this design, because it prevents arbitrary datasets from overriding system or user config (where the latter takes precedence over the former). This include identity information and credentials.

However, if we identify a scenario where this prevents sensible behavior, we will simply stop and reassess the situation -- no biggie.

Reply:
@mih So dist -> dataset -> system -> user?

Reply:
thanks for the lively discussion ;-)  yes, I could have been wrong.  yes, lets proceed with `dist -> dataset -> system -> user`
Cheers!